### DISCUSSION TITLE
Help in chasing down unresolved external symbol

### DISCUSSION THREAD

**Discussion by rowlesmr** [taocpp/PEGTL#310] [Q&A] [answered]
I had some code that was working as single files during testing, and now I have started to consolidate everything into a single application. I am now getting an unresolved external symbol on my read_input function.

As far as I can tell, this is how I've written it, and how it should work:

```
cif_rules.h
#include "tao/pegtl.hpp"
#include "tao/pegtl/contrib/trace.hpp"
namespace pegtl = tao::pegtl;


cif.h
#include "cif_rules.h"
namespace cif {
	template<typename Input> void parse_input(Cif& d, Input&& in);
	template<typename Input> Cif read_input(Input&& in);
}


cif.cpp
#include "cif.h"
namespace cif {
   template<typename Input> void parse_input(Cif& d, Input&& in) {
      pegtl::parse< ::cif::rules::cif, ::cif::rules::Action >(in, d);
   }

   template<typename Input> Cif read_input(Input&& in) {
      Cif cif;
      parse_input(cif, in);
      return cif;
   }
}


main.cpp
#include "cif.h"
pegtl::file_input in("C:\\data\\ideal.cif");
cif::Cif cif = cif::read_input(in);
```

The error is 

`test_main.obj : error LNK2019: unresolved external symbol "struct cif::Cif __cdecl cif::read_input<struct tao::pegtl::file_input<0,struct tao::pegtl::ascii::eol::lf_crlf> &>(struct tao::pegtl::file_input<0,struct tao::pegtl::ascii::eol::lf_crlf> &)" (??$read_input@AEAU?$file_input@$0A@Ulf_crlf@eol@ascii@pegtl@tao@@@pegtl@tao@@@cif@@YA?AUCif@0@AEAU?$file_input@$0A@Ulf_crlf@eol@ascii@pegtl@tao@@@pegtl@tao@@@Z) referenced in function main
`

which I interpret as not being able to find this function:
cif::read_input(tao::pegtl::file_input )

which appears to match what I've defined and declared? Is my reading of this error correct? Any ideas on how to chase this one down? I know this is a little out of your remit...

My full code (a very much work in progress) is available https://github.com/rowlesmr/pdqcif/tree/main/pdqcif/src
I'm using Visual Studio and C++17

**Comment by d-frey** [comment]
In `cif.h` you use `Cif`, but I don't see where it is coming from. Are you missing an include in this header? Are you ignoring any warnings about "Unknown type Cif"?

**Comment by d-frey** [comment, marked as answer]
Actually, I think it's much simpler: You put the definition of a template function into the cpp, but it needs to go to the header. Otherwise it won't get instantiated and therefore the linker won't find it.

**Reply by rowlesmr** [reply]
Ta a lot. After a bunch of buggering around , not helped by namespaces, I got it working. Yes, I had template definitions in the cpp. This is my first multi-file program, so I got bitten.

Thanks for the pointer.

.

and now to figure out what I've changed; it's now failing on files that previously worked...  :X
